server/api/impl/store/apps.go

```diff
@@ -4,94 +4,185 @@
 package store
 
 import (
+	"crypto/sha1" // nolint:gosec
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+
 	"github.com/mattermost/mattermost-plugin-apps/apps"
 	"github.com/mattermost/mattermost-plugin-apps/server/api"
 	"github.com/mattermost/mattermost-plugin-apps/server/utils"
 )
 
-type AppStore struct {
+type appStore struct {
 	*Store
+
+	mutex sync.RWMutex
+
+	installed        map[apps.AppID]*apps.App
+	builtinInstalled map[apps.AppID]*apps.App
 }
 
-var _ api.AppStore = (*AppStore)(nil)
+var _ api.AppStore = (*appStore)(nil)
 
-func newAppStore(st *Store) api.AppStore {
-	s := &AppStore{st}
-	return s
+func (s *appStore) InitBuiltin(builtinApps ...*apps.App) {
+	s.mutex.Lock()
+	if s.builtinInstalled == nil {
+		s.builtinInstalled = map[apps.AppID]*apps.App{}
+	}
+	for _, app := range builtinApps {
+		s.builtinInstalled[app.AppID] = app
+	}
+	s.mutex.Unlock()
 }
 
-func (s AppStore) GetAll() []*apps.App {
-	conf := s.conf.GetConfig()
-	out := []*apps.App{}
-	if len(conf.Apps) == 0 {
-		return out
+func (s *appStore) Configure(conf api.Config) error {
+	newInstalled := map[apps.AppID]*apps.App{}
+
+	for id, key := range conf.InstalledApps {
+		var app *apps.App
+		err := s.mm.KV.Get(api.PrefixInstalledApp+key, &app)
+		if err != nil {
+			s.mm.Log.Error(
+				fmt.Sprintf("failed to load app %s: %s", id, err.Error()))
+		}
+		if app == nil {
+			s.mm.Log.Error(
+				fmt.Sprintf("failed to load app %s: key %s not found", id, api.PrefixInstalledApp+key))
+		}
+
+		newInstalled[apps.AppID(id)] = app
 	}
-	for _, v := range conf.Apps {
-		app := apps.AppFromConfigMap(v)
-		app = s.populateAppWithManifest(app)
-		out = append(out, app)
+
+	s.mutex.Lock()
+	s.installed = newInstalled
+	s.mutex.Unlock()
+
+	return nil
+}
+
+func (s *appStore) Get(appID apps.AppID) (*apps.App, error) {
+	s.mutex.RLock()
+	installed := s.installed
+	builtin := s.builtinInstalled
+	s.mutex.RUnlock()
+
+	app, ok := builtin[appID]
+	if ok {
+		return app, nil
 	}
-	return out
+	app, ok = installed[appID]
+	if ok {
+		return app, nil
+	}
+	return nil, utils.ErrNotFound
 }
 
-func (s AppStore) Get(appID apps.AppID) (*apps.App, error) {
-	conf := s.conf.GetConfig()
-	if len(conf.Apps) == 0 {
-		return nil, utils.ErrNotFound
+func (s *appStore) AsMap() map[apps.AppID]*apps.App {
+	s.mutex.RLock()
+	installed := s.installed
+	builtin := s.builtinInstalled
+	s.mutex.RUnlock()
+
+	out := map[apps.AppID]*apps.App{}
+	for appID, app := range installed {
+		out[appID] = app
 	}
-	v := conf.Apps[string(appID)]
-	if v == nil {
-		return nil, utils.ErrNotFound
+	for appID, app := range builtin {
+		out[appID] = app
 	}
-	app := apps.AppFromConfigMap(v)
-	app = s.populateAppWithManifest(app)
-	return app, nil
+	return out
 }
 
-func (s AppStore) Save(app *apps.App) error {
+func (s *appStore) Save(app *apps.App) error {
```

Comment:
Yes it can, on multiple levels, and no, I don't think it's a big issue - apps and manifests are not very volatile. I'd be more concerned about 2 rapid app installs resulting in only one actually installed, but again, probably not realistic.

Reply:
0/5 Maybe add a comment on that?